lua/ttt2/libraries/migrations.lua

```diff
@@ -0,0 +1,73 @@
+---
+-- A TTT2 version migrations library
+-- @author ZenBre4ker
+-- @module migrations
+
+if SERVER then
+	AddCSLuaFile()
+end
+
+migrations = {}
+migrations.folderPath = "terrortown/migrations/"
+migrations.databaseName = "ttt2_migrations"
+migrations.savingKeys = {
+	executedAt = {
+		typ = "string",
+		default = ""
+	}
+}
+
+sql.CreateSqlTable(migrations.databaseName, migrations.savingKeys)
+migrations.orm = orm.Make(migrations.databaseName)
+
+---
+-- Runs all missing forward gamemode migrations.
+-- @realm shared
+function migrations.Apply()
+	local files = file.Find(migrations.folderPath .. "*.lua", "LUA", "nameasc") or {}
+
+	local fileExecutionCounter = 0
+
+	-- Add Hook to OnLuaError to catch if migrations fail
+	local isSuccess = true
+	local hookName = "OnLuaError"
+	local hookIdentifier = "TTT2MigrationErrorCatcher"
+	hook.Add(hookName, hookIdentifier, function(error, realm, stack, name, id)
+		isSuccess = false
+	end)
+
+	for i = 1, #files do
+		local fileName = files[i]
+		local fullPath = migrations.folderPath .. files[i]
+
+		local fileInfo = migrations.orm:Find(fileName)
+
+		if fileInfo then continue end
+
+		MsgN("[TTT2] Migrating: ", fileName)
+
+		if SERVER then
+			AddCSLuaFile(fullPath)
+		end
+		include(fullPath)
+
+		if not isSuccess then
+			Error("\n[TTT2] Migration failed.\n\n")
```

Comment:
This errors the migration but not the gamemode. You can still play after that. Maybe we should crash our gamemode or should we just ignore it?
It will be erroneous on every server and map restart as the filename doesnt get entered into the database.

Reply:
We can't just run the gamemode after a failed migration.
All code that interacts with whatever the migration was supposed to do expects the migration to be run successfully. If we just run the gamemode after a migration failure we will fail on e.g. table insertions instead or even risk some sort of corruption.